src/AtelierTomato.Markov.Service.Discord/DiscordObjectOIDBuilder.cs

```diff
@@ -1,29 +1,51 @@
 ﻿using AtelierTomato.Markov.Model.ObjectOID;
 using Discord;
+using Microsoft.Extensions.Logging;
 
 namespace AtelierTomato.Markov.Service.Discord
 {
 	public class DiscordObjectOIDBuilder
 	{
-		public static async Task<DiscordObjectOID> Build(IGuild? guild, IChannel channel, ulong messageID, string instance = "discord.com")
+		private readonly ILogger<DiscordObjectOIDBuilder> logger;
+
+		public DiscordObjectOIDBuilder(ILogger<DiscordObjectOIDBuilder> logger)
+		{
+			this.logger = logger;
+		}
+
+		/// <summary>
+		/// Builds a DiscordObjectOID at scope of <see cref="DiscordObjectOID.Channel"/> or <see cref="DiscordObjectOID.Thread"/>.
+		/// </summary>
+		/// <param name="guild">The <see cref="IGuild"/> of the desired DiscordObjectOID. Must contain the <paramref name="channel"/>.</param>
+		/// <param name="channel">The <see cref="IGuildChannel"/> of the desired DiscordObjectOID. Can be <see cref="IThreadChannel"/>. Must be inside the <paramref name="guild"/>.</param>
+		/// <param name="instance">Optional parameter, used if on an alternative instance of Discord.</param>
+		/// <remarks>This uses Discord's API, and thus incurs a big cost.</remarks>
+		/// <returns></returns>
+		public async Task<DiscordObjectOID> Build(IGuild guild, IGuildChannel channel, string instance = "discord.com")
 		{
-			ulong serverID = guild?.Id ?? 0;
-			ulong categoryID, channelID, threadID;
+			if (channel.GuildId != guild.Id)
+			{
+				throw new ArgumentException($"The {nameof(guild)} provided does not contain the {nameof(channel)} provided.", $"{nameof(guild)}, {nameof(channel)}");
+			}
+			ulong categoryID, channelID;
+			ulong? threadID;
 			if (channel is INestedChannel nestedChannel and not IThreadChannel)
 			{
 				categoryID = nestedChannel.CategoryId ?? 0;
 				channelID = channel.Id;
-				threadID = 0;
+				threadID = null;
 			}
 			else if (channel is IThreadChannel threadChannel)
 			{
-				if (guild is not null && await guild.GetChannelAsync(threadChannel.CategoryId!.Value) is not null and INestedChannel parentChannel && parentChannel.CategoryId is not null)
+				if (threadChannel.CategoryId is null)
 				{
-					categoryID = parentChannel.CategoryId.Value;
+					logger.LogWarning("The {ThreadChannel} passed to {Build} did not have a {CategoryId}. This is unexpected.", nameof(IThreadChannel), nameof(Build), nameof(threadChannel.CategoryId));

```

Comment:
for logging don't use template replacements for things that shouldn't be semantic fields